Docs for odo in airgap

boczkowska opened the pull request.

boczkowska commented:
- verified and approved by the QE
- the changes are for 4.2, 4.3, 4.4

cc @bergerhoffer 
Hi Andrea! Please take a look when you have time. 
I have been debating with myself whether I should split  `Pushing the odo init image from the client OS to the disconnected cluster registry` into different modules for different operating systems and for now decided against it since the procedure differs only in some commands. Let me know if you think it would be best to split them.

In both cases (split and unsplit) this piece of documentation looks somewhat bloated. :(

bergerhoffer commented:
@boczkowska Yeah it's a tough call sometimes. Looking at the procedures, I think there are enough steps that differ depending on the OS that it would probably look better to put them in separate modules.

bergerhoffer commented on modules/openshift-developer-cli-pushing-odo-init-image-from-the-client-os-to-the-disconnected-cluster-registry.adoc:
```diff
@@ -0,0 +1,170 @@
+
+* On macOS
++
+.. Restart Docker using Docker UI.
```
using Docker UI -> using the Docker UI?

Same below

bergerhoffer commented on modules/openshift-developer-cli-pushing-odo-init-image-from-the-client-os-to-the-disconnected-cluster-registry.adoc:
```diff
@@ -0,0 +1,170 @@
++
+.. Restart Docker using Docker UI.
++
+.. Run the command:
```
Run the command -> Run the following command

+ other places

bergerhoffer commented on modules/openshift-developer-cli-pushing-odo-init-image-from-the-client-os-to-the-disconnected-cluster-registry.adoc:
```diff
@@ -0,0 +1,170 @@
+$ oc image mirror registry.access.redhat.com/openshiftdo/odo-init-image-rhel7:<tag> <mirror-registry>:5000/openshiftdo/odo-init-image-rhel7:<tag>
+----
+
+. Override the default `odo init` image path by setting an environment variable `ODO_BOOTSTRAPPER_IMAGE`:
```
...by setting the `ODO_BOOTSTRAPPER_IMAGE` environment variable:

Same below

bergerhoffer commented on modules/openshift-developer-cli-mirroring-a-supported-builder-image.adoc:
```diff
@@ -0,0 +1,30 @@
+
+= Mirroring a supported builder image
+
+. Verify that the required images tag are not imported:
```
Should either be "images tags are" or "images tag is"

bergerhoffer commented on modules/openshift-developer-cli-mirroring-a-supported-builder-image.adoc:
```diff
@@ -0,0 +1,30 @@
+$ oc image mirror registry.access.redhat.com/rhscl/nodejs-10-rhel7:<tag> <private_registry>/rhscl/nodejs-10-rhel7:<tag>
+----
+
+. You must periodically re-import the image. Use the --scheduled flag to do it automatically:
```
Is this step importing the image? If so, I would consider rewording a bit.

Maybe just say "Import the image" as the step.  And then have a callout on that line that says something like "You must periodically re-import the image. The `--scheduled` flag re-imports the image automatically." 

Not sure if you typically use callouts in these sections or not, but it doesn't have to be a callout.

bergerhoffer commented on modules/openshift-developer-cli-overwriting-a-mirror-registry.adoc:
```diff
@@ -0,0 +1,22 @@
+
+= Overwriting the mirror registry in {odo-title}
+
+To be able to download npm packages for Node.js dependencies and Maven packages Java dependencies from a private mirror registry, you must create and configure a mirror npm or Maven registry on the cluster. You can then overwrite the mirror registry on an existing component or when creating a new component.
```
To be able to -> To

bergerhoffer commented on modules/openshift-developer-cli-creating-and-deploying-a-nodejs-application-with-odo.adoc:
```diff
@@ -1,6 +1,7 @@
 //
-// * cli_reference/openshift_developer_cli/creating-a-single-component-application-with-odo    
+// * cli_reference/openshift_developer_cli/creating-a-single-component-application-with-odo
+// * modules/openshift-developer-cli-creating-and-pushing-a-component-from-client-os-to-the-restricted-cluster
```
Again, shouldn't be including a module from another module.

boczkowska replied:
Restructured

boczkowska commented on cli_reference/openshift_developer_cli/using-odo-in-a-restricted-environment.adoc:
```diff
@@ -0,0 +1,18 @@
+:context: using-odo-in-a-restricted-environment
+toc::[]
+
+To run {odo-title} in a disconnected {product-title} cluster or a cluster provisioned in a restricted environment, you need to push the `odo init` image to the private registry of the cluster or the default internal registry, and then to overwrite the `odo init` image path using the `ODO_BOOTSTRAPPER_IMAGE` environment variable.
```
@bergerhoffer @kalexand-rh 

Regarding "disconnected" - I tried to follow the pattern introduced in this doc https://docs.openshift.com/container-platform/4.3/operators/olm-restricted-networks.html where network or environment are referred to as "restricted" but a cluster is referred to as "disconnected".

boczkowska commented on cli_reference/openshift_developer_cli/using-odo-in-a-restricted-environment.adoc:
```diff
@@ -0,0 +1,18 @@
+* Create and configure a mirror of `npm` and `maven` private registries with your a disconnected cluster. See xref:../../installing/install_config/installing-restricted-networks-preparations.adoc[Creating a mirror registry for installation in a restricted network] for more information.
+
+include::modules/openshift-developer-cli-pushing-odo-init-image-from-the-client-os-to-the-disconnected-cluster-registry.adoc[leveloffset=+1]
+include::cli_reference/openshift_developer_cli/odo-creating-and-pushing-a-component-from-client-os-to-the-disconnected-cluster.adoc[leveloffset=+1]
```
Restructured.

boczkowska commented on modules/openshift-developer-cli-pushing-odo-init-image-from-the-client-os-to-the-disconnected-cluster-registry.adoc:
```diff
@@ -0,0 +1,170 @@
+
+= Pushing the `odo init` image from the client OS to the disconnected cluster registry
+
+== For a cluster that has a mirror registry
```
Thank you! Did exactly that and broke the OS-related instructions into separate modules.

boczkowska commented on modules/openshift-developer-cli-pushing-odo-init-image-from-the-client-os-to-the-disconnected-cluster-registry.adoc:
```diff
@@ -0,0 +1,170 @@
+
+. Trust a CA in your client platform and log into the {product-title} mirror registry:
+
+* On Linux
```
I don't think it would be correct. Linux has many distributions, odo can be used on many of them.

Preeticp commented on modules/openshift-developer-cli-mirroring-a-supported-builder-image.adoc:
```diff
@@ -0,0 +1,35 @@
+
+
+
+.Procedure
```
The procedure should come after the heading. But strangely I see the preview reflecting it correctly. 
Also, the [id="<module title>_{context}"] is missing.

Preeticp commented on modules/openshift-developer-cli-overwriting-a-mirror-registry.adoc:
```diff
@@ -0,0 +1,22 @@
+//
+// *modules/openshift-developer-cli-creating-and-pushing-a-component-from-client-os-to-the-disconnected-cluster
+
+= Overwriting the mirror registry in {odo-title}
```
The [id="<module title>_{context}"] is missing.

Preeticp commented on modules/openshift-developer-cli-mirroring-a-supported-builder-image.adoc:
```diff
@@ -0,0 +1,35 @@
+$ oc tag <mirror-registry>:<port>/rhscl/nodejs-10-rhel7:<tag> nodejs-10-rhel7:latest --scheduled <1>
+----
++
+<1> You must periodically re-import the image. The `--scheduled` flag enables automatic re-import of the image.
```
I think you can simply add this information as a sentence and remove the callout. IMHO, If it was a huge block of code and user would be confused about what you are referring too then we could use call outs.

Preeticp commented on modules/openshift-developer-cli-mirroring-a-supported-builder-image.adoc:
```diff
@@ -0,0 +1,35 @@
+<1> You must periodically re-import the image. The `--scheduled` flag enables automatic re-import of the image.
+
+. Verify the images with the given tag have been imported:
++
```
Verify that the...

Preeticp commented on modules/openshift-developer-cli-overwriting-a-mirror-registry.adoc:
```diff
@@ -0,0 +1,22 @@
+// Module included in the following assemblies:
+//
+// *modules/openshift-developer-cli-creating-and-pushing-a-component-from-client-os-to-the-disconnected-cluster
```
Replace with: // cli_reference/openshift_developer_cli/using_odo_in_a_restricted_environment/creating-and-deploying-a-component-to-the-disconnected-cluster

Preeticp commented on modules/openshift-developer-cli-overwriting-a-mirror-registry.adoc:
```diff
@@ -0,0 +1,22 @@
+
+= Overwriting the mirror registry in {odo-title}
+
+To download npm packages for Node.js dependencies and Maven packages Java dependencies from a private mirror registry, you must create and configure a mirror npm or Maven registry on the cluster. You can then overwrite the mirror registry on an existing component or when creating a new component.
```
Maven packages for Java dependencies
or when you create a new component

Preeticp commented on modules/openshift-developer-cli-pushing-the-init-image-directly-on-linux.adoc:
```diff
@@ -0,0 +1,67 @@
+$ docker push <registry_path>/openshiftdo/odo-init-image-rhel7:<tag>
+----
+
+. Override the default `odo init` image path by setting the environment variable `ODO_BOOTSTRAPPER_IMAGE`:
```
by setting the `ODO_BOOTSTRAPPER_IMAGE` environment variable

Preeticp commented on modules/openshift-developer-cli-pushing-the-init-image-directly-on-macos.adoc:
```diff
@@ -0,0 +1,67 @@
+$ docker push <registry_path>/openshiftdo/odo-init-image-rhel7:<tag>
+----
+
+. Override the default `odo init` image path by setting the environment variable `ODO_BOOTSTRAPPER_IMAGE`:
```
by setting the `ODO_BOOTSTRAPPER_IMAGE` environment variable

Preeticp commented on modules/openshift-developer-cli-pushing-the-init-image-directly-on-windows.adoc:
```diff
@@ -0,0 +1,68 @@
+PS C:\> echo <tls.crt> | base64 -d > ca.crt
+----
+
+. Trust a CA in your client platform/ Execute the following command in PowerShell as administrator:
```
As an administrator, trust a CA in your client platform by executing the following command in PowerShell:

Preeticp commented on modules/openshift-developer-cli-pushing-the-init-image-directly-on-windows.adoc:
```diff
@@ -0,0 +1,68 @@
+PS C:\> docker push <registry_path>/openshiftdo/odo-init-image-rhel7:<tag>
+----
+
+. Override the default `odo init` image path by setting the environment variable `ODO_BOOTSTRAPPER_IMAGE`:
```
by setting the `ODO_BOOTSTRAPPER_IMAGE` environment variable:

Preeticp commented on modules/openshift-developer-cli-pushing-the-init-image-to-a-mirror-registry-on-linux.adoc:
```diff
@@ -0,0 +1,40 @@
+
+.Procedure
+
+. Use `base64` to encode your mirror registry's root certification authority (CA) content:
```
to encode the root certification authority (CA) content of your mirror registry
Avoid anthropomorphism.

Preeticp commented on modules/openshift-developer-cli-pushing-the-init-image-to-a-mirror-registry-on-linux.adoc:
```diff
@@ -0,0 +1,40 @@
+$ oc image mirror registry.access.redhat.com/openshiftdo/odo-init-image-rhel7:<tag> <mirror-registry>:5000/openshiftdo/odo-init-image-rhel7:<tag>
+----
+
+. Override the default `odo init` image path by setting an environment variable `ODO_BOOTSTRAPPER_IMAGE`:
```
setting the `ODO_BOOTSTRAPPER_IMAGE` environment variable

Preeticp replied:
Please make this change in all places applicable.

Preeticp commented on modules/openshift-developer-cli-pushing-the-init-image-to-a-mirror-registry-on-macos.adoc:
```diff
@@ -0,0 +1,37 @@
+
+.Procedure
+
+. Use `base64` to encode your mirror registry's root certification authority (CA) content:
```
to encode the root certification authority (CA) content of your mirror registry
Please make this change where ever applicable.

Preeticp commented on modules/openshift-developer-cli-pushing-the-init-image-to-a-mirror-registry-on-macos.adoc:
```diff
@@ -0,0 +1,37 @@
+
+. Copy the encoded root CA certificate to the appropriate location:
++
+.. Restart Docker using Docker UI.
```
using the Docker

Preeticp commented on modules/openshift-developer-cli-pushing-the-init-image-to-a-mirror-registry-on-windows.adoc:
```diff
@@ -0,0 +1,43 @@
+PS C:\> echo <content_of_additional_ca> | base64 -d > disconnect-ca.crt
+----
+
+. Copy the encoded root CA certificate to the appropriate location. Run PowerShell as administrator and run the following command:
```
As an administrator, copy...location by executing the following command in PowerShell:

boczkowska commented:
@bergerhoffer @Preeticp 
Hi guys! I have addressed your comments. Thank you so much!

The docs have also been restructured. Please take a look when you have time and let me know if I missed anything.

Preeticp commented on modules/openshift-developer-cli-pushing-the-odo-init-image-to-a-mirror-registry.adoc:
```diff
@@ -0,0 +1,119 @@
+[id="pushing-the-odo-init-image-to-a-mirror-registry_{context}"]
+= Pushing the `odo init` image to a mirror registry
+
+Push the `odo init` image to a cluster with a mirror registry as follows:
```
Suggest adding, Depending on your operating system, you can push...

boczkowska replied:
Thank you. Changed. 

Also changed wording of a phrase about `ODO_BOOTSTRAPPER_IMAGE` throughout the docs.

bergerhoffer commented on _topic_map.yml:
```diff
@@ -1207,6 +1207,13 @@ Topics:
+  - Name: Using odo in a restricted environment  
+    Dir: using_odo_in_a_restricted_environment
+    Topics: 
+    - Name: Pushing the `odo init` image to the restricted cluster registry
```
The backticks are rendering in the TOC. You should remove them because I don't think they're accepted in the TOC.

bergerhoffer commented on cli_reference/openshift_developer_cli/using_odo_in_a_restricted_environment/creating-and-deploying-a-component-to-the-disconnected-cluster.adoc:
```diff
@@ -0,0 +1,18 @@
+* Install `oc` on the client operating system.
+* Install `{odo-title}` on the client operating system.
+* Create an {product-title} cluster in a restricted environment.
+* Create and configure a mirror of npm and Maven private registries within your disconnected cluster. 
```
a mirror of -> a mirror of the

bergerhoffer commented on modules/openshift-developer-cli-pushing-the-odo-init-image-to-a-mirror-registry.adoc:
```diff
@@ -0,0 +1,119 @@
+PS C:\> echo <content_of_additional_ca> | base64 -d > disconnect-ca.crt
+----
+
+. As an administrator, copy the encoded root CA certificate to the appropriate location by executing the following command in PowerShell:
```
Either remove "in PowerShell" completely (because you already executed a command in it in the previous step), or move the "in PowerShell" to step 1 if you think it's important to call out. 

Same for the other procedure for Windows

bergerhoffer commented on cli_reference/openshift_developer_cli/using_odo_in_a_restricted_environment/creating-and-deploying-a-component-to-the-disconnected-cluster.adoc:
```diff
@@ -0,0 +1,18 @@
+include::modules/common-attributes.adoc[]
+:context: creating-and-deploying-a-component-to-the-disconnected-cluster
+toc::[]
+
```
I really do think that this restructure is an improvement, breaking the assemblies out. However, now this assembly (and most of its procedures) don't have any lead-in information. Suggest adding at least at the top of the assembly, some information on what they're doing here/why.

bergerhoffer commented on modules/openshift-developer-cli-mirroring-a-supported-builder-image.adoc:
```diff
@@ -0,0 +1,34 @@
+. Verify that the required images tag is not imported:
++
+----
+$ oc describe is nodejs -n openshift
```
Not critical, but probably would be a good idea to tell users what to expect the output of this command to be. And what should they do if it is important? Should they stop following the procedure?

bergerhoffer commented on modules/openshift-developer-cli-mirroring-a-supported-builder-image.adoc:
```diff
@@ -0,0 +1,34 @@
+. Verify that the images with the given tag have been imported:
++
+----
+$ oc describe is nodejs -n openshift
```
Same comment as above about telling users what they should expect to see here.

amitkrout commented:
@boczkowska Let me know if you need info from my end

boczkowska commented:
@bergerhoffer  after some discussion with @Preeticp , we added another module - `About odo in a restricted environment` in which we added more information for users and some useful references. 

I've also fixed all of the errors you pointed to, added commands' outputs and introductions. Please take a look when you have time.

bergerhoffer commented on cli_reference/openshift_developer_cli/using_odo_in_a_restricted_environment/about-odo-in-a-restricted-environment.adoc:
```diff
@@ -0,0 +1,19 @@
+To start working in a disconnected cluster, you must first xref:../../../cli_reference/openshift_developer_cli/using_odo_in_a_restricted_environment/pushing-the-odo-init-image-to-the-restricted-cluster-registry.adoc#pushing-the-odo-init-image-to-a-mirror-registry_pushing-the-odo-init-image-to-the-restricted-cluster-registry[push the `odo init` image to the registry of the cluster] and then overwr ...
+
+
+After you push the `odo init image`, you must xref:../../../cli_reference/openshift_developer_cli/using_odo_in_a_restricted_environment/creating-and-deploying-a-component-to-the-disconnected-cluster.adoc#mirroring-a-supported-builder-image_creating-and-deploying-a-component-to-the-disconnected-cluster[mirror a supported builder image] from the registry, xref:../../../cli_reference/openshift_devel ...
```
This is incomplete: `A builder image is necessary to configure a runtime environment for your application, it also contains ...`.

Suggest `and also contains`

bergerhoffer commented on modules/openshift-developer-cli-pushing-the-odo-init-image-to-a-mirror-registry.adoc:
```diff
@@ -0,0 +1,119 @@
+PS C:\> echo <content_of_additional_ca> | base64 -d > disconnect-ca.crt
+----
+
+. As an administrator, copy the encoded root CA certificate to the appropriate location byt executing the following command:
```
byt -> by

bergerhoffer commented on cli_reference/openshift_developer_cli/using_odo_in_a_restricted_environment/about-odo-in-a-restricted-environment.adoc:
```diff
@@ -0,0 +1,19 @@
+[id="about-odo-in-a-restricted-environemnt"]
+= About {odo-title} in a restricted environemnt
```
environemnt -> environment

(Should fix in anchor and context too)

boczkowska commented:
@bergerhoffer fixed! Thank you!

bergerhoffer commented:
Thanks for the updates @boczkowska! LGTM now, merging!

bergerhoffer commented:
/cherrypick enterprise-4.4

bergerhoffer commented:
/cherrypick enterprise-4.3

bergerhoffer commented:
/cherrypick enterprise-4.2
